server/src/main/java/org/elasticsearch/index/query/RangeQueryBuilder.java

```diff
@@ -436,11 +436,11 @@ public String getWriteableName() {
     protected MappedFieldType.Relation getRelation(final CoordinatorRewriteContext coordinatorRewriteContext) {
         final MappedFieldType fieldType = coordinatorRewriteContext.getFieldType(fieldName);
         if (fieldType instanceof final DateFieldMapper.DateFieldType dateFieldType) {
-            if (coordinatorRewriteContext.hasTimestampData() == false) {
+            if (coordinatorRewriteContext.hasTimestampData(fieldName) == false) {
```

Comment:
do range queries on date fields that are not `@timestamp` or `event.ingested` still work? I see that `hasTimestampData` throws exception? 

On a similar note, I wonder if those new methods exposed in the context class are too broad in that they take a generic field name string as an argument, but after all support only two possible values for it. Should we then have two explicit separate methods for these? That though would require moving the if to the caller code. I don't have a strong opinion on this, but it feels like at least the getMin and getMax should not throw IAE and return some placeholder like before for other date fields and the previous logic still works as before?

Reply:
Once the refactoring for removing the `Map<String, DateFieldMapper.DateFieldType>` was done (in [this commit](https://github.com/elastic/elasticsearch/pull/110352/commits/005ed5035080791e91baaff2f0313123830b10f2)), I then started work to change this class, but I ended up not making any changes here as I think it is coded right. 

First, the `getRelation` call is only called in during a coordinator rewrite. 

Second `coordinatorRewriteContext.getFieldType` returns `null` if you pass in a fieldName that is not supported for coordinator rewrites. If you get back null, the `if` check on the next line will return false and none of the rest of the code executes - we just return `MappedFieldType.Relation.INTERSECTS`.

The other calls (which can throw IllegalArgumentException now) are all guarded by that if-check.

It also feels wrong to put separate `getMin` and `getMax` methods for `@timestamp` and `event.ingested` because that means the RangeQueryBuilder has to know (and hardcode) knowledge of what timestamps are in cluster state (law of Demeter violation). The `RangeQueryBuilder` just knows what `fieldName` it is querying against and that's all it needs to know. It asks the `coordinatorRewriteContext` whether it supports that field and has useful min/max ranges to return.

Reply:
I agree. I missed that getFieldType returns null . You could slightly adjust `CoordinatorRewriteContext#getFieldType` to return a `DateFieldType` and replace the instanceof with a null check, to make this all more explicit. it is though odd that we return the field type, but we never really do anything with it. Up to you how you want to clean this up a little.